src/dvc_objects/fs/generic.py

```diff
@@ -397,3 +409,118 @@ def test_links(
         from_fs.remove(from_file)
 
     return ret
+
+
+def status(
+    fs: "FileSystem",
+    file_paths: Union["AnyFSPath", Collection["AnyFSPath"]],
+    callback: "Callback" = DEFAULT_CALLBACK,
+    batch_size: Optional[int] = None,
+) -> Dict[str, bool]:
+    """Return status (existence) of files in fs.
+
+    Runs batched fs.exists() calls in parallel with fs.ls() until all paths
+    have been checked.
+    """
+    if not file_paths:
+        return {}
+    paths = {file_paths} if isinstance(file_paths, str) else set(file_paths)
+    if len(paths) == 1:
+        path = paths.pop()
+        return {path: fs.exists(path)}
+
+    paths_lock = threading.Lock()
+    results: Dict[str, bool] = {}
+    results_lock = threading.Lock()
+    callback.set_size(len(paths))
+    jobs = batch_size or fs.jobs
+    exists_jobs = jobs - 1 if jobs > 1 else 1
+    executor = ThreadPoolExecutor(max_workers=2, cancel_on_error=True)
+    logger.debug("Querying status for '%d' files", len(paths))
+    exist_fut = executor.submit(
+        _exist_query,
+        fs,
+        paths,
+        paths_lock,
+        results,
+        results_lock,
+        exists_jobs,
+        callback,
+    )
+    list_fut = executor.submit(
+        _list_query,
+        fs,
+        paths,
+        paths_lock,
+        results,
+        results_lock,
+        callback,
+    )
+    _done, not_done = futures.wait(
+        [exist_fut, list_fut], return_when=futures.FIRST_COMPLETED
+    )
+    for fut in not_done:
+        fut.cancel()
+    # NOTE: if we started a long running lsdir it will continue to run in
+    # the background until the task completes
```

Comment:
The main issue here is that fsspec ls() (or find) only returns a complete result (instead of using iterators). So there's no straightforward way for us to short-circuit the fsspec ls call once it has been started. So if we call ls() on what ends up being a large directory with a large # of object versions, that call will keep running in the background thread until fsspec has gotten the entire listing and returned it.

Reply:
note check iterdir: https://github.com/fsspec/s3fs/blob/main/s3fs/core.py#L703